client/consensus/babe/src/lib.rs

```diff
@@ -521,14 +531,87 @@ where
 	let (worker_tx, worker_rx) = channel(HANDLE_BUFFER_SIZE);
 
 	let answer_requests =
-		answer_requests(worker_rx, config.0, client, babe_link.epoch_changes.clone());
+		answer_requests(worker_rx, config.0, client.clone(), babe_link.epoch_changes.clone());
+
+	let finality_notifications = client.finality_notification_stream();
+	let clean_weights = aux_storage_cleanup(client, finality_notifications);
+
+	// TODO: PROPOSAL
+	// The inner futures are "never-ending" tasks.
+	// Maybe we should replace `join` with `select` to exit from the worker as
+	// soon as one of its inner futures terminates.
+	// The `join` will keep polling the others even if one of them has finished.
 	Ok(BabeWorker {
-		inner: Box::pin(future::join(inner, answer_requests).map(|_| ())),
+		inner: Box::pin(future::join3(inner, answer_requests, clean_weights).map(|_| ())),
 		slot_notification_sinks,
 		handle: BabeWorkerHandle(worker_tx),
 	})
 }
 
+// Remove obsolete block's weight data by leveraging finality notifications.
+// This includes data for all finalized blocks (excluded the most recent one)
+// and all stale branches.
+async fn aux_storage_cleanup<B: BlockT, C>(
+	client: Arc<C>,
+	mut finality_notifications: FinalityNotifications<B>,
+) where
+	C: HeaderMetadata<B> + AuxStore,
+{
+	while let Some(notification) = finality_notifications.next().await {
+		let mut aux_keys = HashSet::new();
+		let mut height_limit = Zero::zero();
+
+		// Cleans data for finalized block's ancestors down to, and including, the previously
+		// finalized one.
+
+		let first_new_finalized = notification.tree_route.get(0).unwrap_or(&notification.hash);
+
+		match client.header_metadata(*first_new_finalized) {
+			Ok(meta) => {
+				aux_keys.insert(aux_schema::block_weight_key(meta.parent));
+				height_limit = meta.number.saturating_sub(One::one());
+			},
+			Err(err) => {
+				warn!(target: "babe", "header lookup fail while cleaning data for block {}: {}", first_new_finalized.to_string(), err.to_string());
+			},
+		}
+
+		for hash in notification.tree_route.iter() {
+			aux_keys.insert(aux_schema::block_weight_key(hash));
+		}
+
+		// Cleans data for stale branches.
+
+		for head in notification.stale_heads.iter() {
+			let mut hash = *head;
+			// Insert stale blocks hashes until canonical chain is not reached.
+			// Soon or late we should hit an element already present within the `aux_keys` set.
+			while aux_keys.insert(aux_schema::block_weight_key(hash)) {
+				match client.header_metadata(hash) {
+					Ok(meta) => {
+						// A fallback in case of malformed notification.
+						// This should never happen and must be considered a bug.
+						if meta.number.le(&height_limit) {
```

Comment:
```suggestion
						if meta.number <= &height_limit {
```

Can we not do this?  (makes it a little bit easier to read)